optuna/study/_tell.py

```diff
@@ -22,59 +20,96 @@
+            raise ValueError(
+                "No values were told. Values are required when state is TrialState.COMPLETE."
+            )
+    elif state in (TrialState.PRUNED, TrialState.FAIL):
+        if values is not None:
+            raise ValueError(
+                "Values were told. Values cannot be specified when state is "
+                "TrialState.PRUNED or TrialState.FAIL."
+            )
+    elif state is not None:
+        raise ValueError(f"Cannot tell with state {state}.")
+
+
+def _check_values(
+    study: "optuna.Study", value_or_values: Union[float, Sequence[float]]
+) -> Optional[str]:
+    if isinstance(value_or_values, Sequence):
+        values = value_or_values
     else:
-        _original_values = [original_value]
+        values = [value_or_values]
 
-    if n_objectives != len(_original_values):
+    if len(study.directions) != len(values):
         return (
-            None,
-            (
-                f"The number of the values "
-                f"{len(_original_values)} did not match the number of the objectives "
-                f"{n_objectives}."
-            ),
+            f"The number of the values {len(values)} did not match the number of the objectives "
+            f"{len(study.directions)}."
         )
 
-    _checked_values = []
-    for v in _original_values:
-        checked_v, failure_message = _check_single_value(v, trial_number)
+    for v in values:
+        failure_message = _check_single_value(v)
         if failure_message is not None:
             # TODO(Imamura): Construct error message taking into account all values and do not
-            #  early return
-            # `value` is assumed to be ignored on failure so we can set it to any value.
-            return None, failure_message
-        elif isinstance(checked_v, float):
-            _checked_values.append(checked_v)
-        else:
-            assert False
-
-    return _checked_values, None
+            # early return `value` is assumed to be ignored on failure so we can set it to any
+            # value.
+            return failure_message
 
+    return None
 
-def _check_single_value(
-    original_value: Optional[float], trial_number: int
-) -> Tuple[Optional[float], Optional[str]]:
-    value = None
-    failure_message = None
 
+def _check_single_value(value: float) -> Optional[str]:
     try:
-        value = float(original_value)  # type: ignore
-    except (
-        ValueError,
-        TypeError,
-    ):
-        failure_message = f"The value {repr(original_value)} could not be cast to float."
+        float(value)
+    except (ValueError, TypeError):
+        return f"The value {repr(value)} could not be cast to float."
 
-    if value is not None and math.isnan(value):
-        value = None
-        failure_message = f"The value {original_value} is not acceptable."
+    if math.isnan(value):
+        return f"The value {value} is not acceptable."
 
-    return value, failure_message
+    return None
```

Comment:
[Nit] Not a strong opinion, but I think this function is simple enough to be directly put inside `_check_values`, which would make the role of each function clearer. What do you think?